compiler/src/language_server/hints/mod.rs

```diff
@@ -0,0 +1,198 @@
+            if let Some(input) = constant_evaluator.run(&db) {
+                fuzzer.update_input(
+                    &db,
+                    input.clone(),
+                    constant_evaluator.get_fuzzable_closures(&input),
+                );
+                break 'new_insight Some(input);
+            }
+            if let Some(input) = fuzzer.run(&db) {
+                log::warn!("Fuzzer found a problem!");
+                break 'new_insight Some(input);
+            }
+            None
+        };
+
+        if let Some(input) = input_with_new_insight {
+            let hints = constant_evaluator
+                .get_hints(&db, &input)
+                .into_iter()
+                // The fuzzer returns groups of related hints.
+                .map(|hint| vec![hint])
+                .chain(fuzzer.get_hints(&db, &input).into_iter())
+                // Make hints look like comments.
+                .map(|mut hint_group| {
+                    for hint in &mut hint_group {
+                        hint.text = format!("{}# {}", quasi_spaces(2), hint.text);
+                    }
+                    hint_group
+                })
+                // Show related hints at the same indentation.
+                .map(|mut hint_group| {
+                    hint_group.align_hint_columns();
+                    hint_group
+                })
+                .flatten()
+                .sorted_by_key(|hint| hint.position)
+                .collect_vec();
+
+            if let Some(path) = input.to_path() {
+                let hints_file = path.clone_with_extension("candy.hints");
+                let content = hints.iter().map(|hint| format!("{hint:?}")).join("\n");
+                fs::write(hints_file.clone(), content).unwrap();
+            }
+
+            // Only show the most important hint per line.
+            let hints = hints
+                .into_iter()
+                .group_by(|hint| hint.position.line)
+                .into_iter()
+                .map(|(_, hints)| hints.max_by_key(|hint| hint.kind).unwrap())
+                .collect_vec();
+
+            outgoing_hints.report_hints(input, hints).await;
+        }
+    }
+}
+
+struct OutgoingHints {
+    sender: Sender<(Input, Vec<Hint>)>,
+    last_sent: HashMap<Input, Vec<Hint>>,
+}
+impl OutgoingHints {
+    fn new(sender: Sender<(Input, Vec<Hint>)>) -> Self {
+        Self {
+            sender,
+            last_sent: HashMap::new(),
+        }
+    }
+
+    async fn report_hints(&mut self, input: Input, hints: Vec<Hint>) {
+        if self.last_sent.get(&input) != Some(&hints) {
+            self.last_sent.insert(input.clone(), hints.clone());
+            self.sender.send((input, hints)).await.unwrap();
+        }
+    }
+}
+
+/// VSCode trims multiple leading spaces to one. That's why we use an
+/// [em quad](https://en.wikipedia.org/wiki/Quad_(typography)) instead, which
+/// per definition has the same width as a normal space.
```

Comment:
That's not quite its definition: “An em quad is a space that is one [em](https://en.wikipedia.org/wiki/Em_(typography)) wide; as wide as the height of the font.” In fact, that is usually wider than one space, as is also visualized on this handy page: https://jkorpela.fi/chars/spaces.html. For monospace fonts it might work, but not in general.